src/passes/CastRefining.cpp

```diff
@@ -0,0 +1,292 @@
+    // can affect needed types. For example, imagine that we replace all $A with
+    // $B, and we had types like this:
+    //
+    //  $C = [.., $A, ..]
+    //  $D = [.., $B, ..]
+    //
+    // After replacing $A with $B, we cause $C and $D to be structurally
+    // identical. However, we will not actually merge those in the normal course
+    // of optimization: the type rewriter will create a single new rec group for
+    // all new types anyhow, so they all remain distinct from each other. The
+    // only thing that can merge them is if we run TypeMerging, which is not run
+    // by default exactly for this reason, that it can limit optimizations.
+    // Thus, this pass does only "safe" merging, that cannot limit later
+    // optimizations.
+
+    TypeMapper::TypeUpdates mapping;
+
+    for (auto type : subTypes.types) {
+      if (!type.isStruct()) {
+        // TODO: support arrays and funcs
+        continue;
+      }
+
+      // Add a mapping of types that are never created (and none of their
+      // subtypes) to the bottom type. This is valid because all locations of
+      // that type, like a local variable, will only contain null at runtime.
+      // Likewise, if we have a ref.test of such a type, we can only be looking
+      // for a null at best. This can be seen as "refining" uses of these
+      // never-created types to the bottom type.
+      //
+      // We check this first as it is the most powerful change.
+      if (createdTypesOrSubTypes.count(type) == 0) {
+        mapping[type] = type.getBottom();
+        continue;
+      }
+
+      // Otherwise, apply a refining if we found one before.
+      if (auto iter = refinableTypes.find(type); iter != refinableTypes.end()) {
+        mapping[type] = iter->second;
+      }
+    }
+
+    if (mapping.empty()) {
+      return;
+    }
+
+    // A TypeMapper that handles the patterns we have in our mapping, where we
+    // end up mapping a type to a *subtype*. We need to properly create
+    // supertypes while doing this rewriting. For example, say we have this:
+    //
+    //  A :> B :> C
+    //
+    // Say we see B is never created, so we want to map B to its subtype C. C's
+    // supertype must now be A.
+    class CastRefiningTypeMapper : public TypeMapper {
+    public:
+      CastRefiningTypeMapper(Module& wasm, const TypeUpdates& mapping)
+        : TypeMapper(wasm, mapping) {}
+
+      std::optional<HeapType> getSuperType(HeapType oldType) override {
+        auto super = oldType.getSuperType();
+        if (!super) {
+          return super;
+        }
+
+        // Go up the chain of supertypes, skipping things we are mapping away,
+        // as those things will not appear in the output. This skips B in the
+        // example above.
+        while (1) {
+          if (mapping.count(*super)) {
+            super = super->getSuperType();
+          } else {
+            return super;
+          }
+        }
```

Comment:
This should check that the new `super` exists before accessing it (and it would be good to add a test that exercises this). It can also be a two-liner:
```suggestion
        for (; super && mapping.count(*super); super = super->getSuperType()) {}
        return super;
```

Reply:
Good catch. This was actually already tested, but I guess nothing bad happened to happen from dereferencing the optional (whatever we found there wasn't in the mapping, so it worked out... Fixed.